src/sql/src/plan/lowering.rs

```diff
@@ -178,22 +201,94 @@ impl HirRelationExpr {
                         typ,
                     })
                 }
-                Get { id, typ } => {
-                    // Get statements are only to external sources, and are not correlated with `get_outer`.
-                    get_outer.product(SR::Get { id, typ })
+                Get { id, typ } => match id {
+                    expr::Id::Local(local_id) => {
+                        let cte_desc = cte_map.get(&local_id).unwrap();
+                        let get_cte = SR::Get {
+                            id: expr::Id::Local(cte_desc.new_id.clone()),
+                            typ: cte_desc.relation_type.clone(),
+                        };
+                        if get_outer == cte_desc.outer_relation {
+                            // If the CTE was applied to the same exact relation, we can safely
+                            // return a `Get` relation.
+                            get_cte
+                        } else {
+                            // Otherwise, the new outer relation may contain more columns from some
+                            // intermediate scope placed between the definition of the CTE and the
+                            // reference of the CTE. In that case, `get_outer` is guaranteed to
+                            // contain the columns of the outer relation the CTE was applied to at
+                            // its prefix. Since, we must return a relation containing `get_outer`'s
+                            // column at the beginning, we must build a join between `get_outer` and
+                            // `get_cte` on their common columns.
+                            let mut equivalences = Vec::new();
+                            let oa = get_outer.arity();
+                            let cte_outer_columns = cte_desc.relation_type.arity() - typ.arity();
+                            for position in 0..cte_outer_columns {
+                                equivalences.push(vec![
+                                    expr::MirScalarExpr::column(position),
+                                    expr::MirScalarExpr::column(oa + position),
+                                ]);
+                            }
+
+                            // Project out the second copy of the common between `get_outer` and
+                            // `cte_desc.outer_relation`.
+                            let projection = (0..oa)
+                                .chain(oa + cte_outer_columns..oa + cte_outer_columns + typ.arity())
+                                .collect_vec();
+                            SR::join_scalars(vec![get_outer, get_cte], equivalences)
+                                .project(projection)
+                        }
+                    }
+
+                    _ => get_outer.product(SR::Get { id, typ }),
```

Comment:
We should preserve the original comment above this line that says a get reference to an external source cannot be correlated with `get_outer`.